### PR TITLE
Travel Logic

### LINKED ISSUE #26: Writing Travel Logic

### PR DESCRIPTION
**Pull request by ColonelKai**

### CODE REVIEW

File: src/main/java/org/colonelkai/publictransit/TravelManager.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    public Optional<Travel> getTravelFromPlayer(User player) {
+        return this.getTravelFromPlayer(player.getUniqueId());
+    }
+
+    private Optional<Travel> getTravelFromScheduler(Scheduler scheduler) {
+        return this.travelMap.entrySet()
+                .parallelStream()
+                .filter((e) -> {return e.getValue().equals(scheduler);})
```
Comment (mosemister) [inline src/main/java/org/colonelkai/publictransit/TravelManager.java]: Single line body (either make it a multi-lined body or a single line lamda)

multi-lined body
```java
filter(e -> {
  return e.getValue().equals(scheduler);
})
```

single-lined lamda

```java
filter(e -> e.getValue().equals(scheduler))
```

File: src/main/java/org/colonelkai/publictransit/TravelManager.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    private Scheduler scheduleNextTravel(Travel travel) {
+        return TranslateCore.getScheduleManager().schedule()
+                .setDelay(travel.getCurrentNode().getTime())
+
+                // RUNNER
+                .setRunner(
+                        (Scheduler s) -> {
+                            // check if we're still on the list, or if we've been removed.
+                            Optional<Travel> optionalTravel = PublicTransit.getPlugin().getTravelManager().getTravelFromScheduler(s);
```
Comment (mosemister) [inline src/main/java/org/colonelkai/publictransit/TravelManager.java]: All this is a duplicate of your field ``consumer``

File: src/main/java/org/colonelkai/publictransit/TravelManager.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                        }
+
+                // RUN AFTER-ER
+                ).setToRunAfter(
+                       TranslateCore.getScheduleManager().schedule()
+                               .setDelay(travel.getCurrentNode().getTime())
+                               .setRunner(scheduler -> {
+                                   Scheduler newScheduler = this.scheduleNextTravel(travel);
+                                   PublicTransit.getPlugin().getTravelManager().updateScheduler(
+                                           travel, newScheduler
```
Comment (mosemister) [inline src/main/java/org/colonelkai/publictransit/TravelManager.java]: This is horrible. Try

```java
PublicTransit
  .getPlugin()
  .getTravelManager()
  .updateScheduler(travel, newScheduler);
```

File: src/main/java/org/colonelkai/publictransit/TravelManager.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+public class TravelManager {
+    private final HashMap<Travel, Scheduler> travelMap = new HashMap<>();
+
+    private final Consumer<Scheduler> consumer = (Scheduler s) -> {
+        // check if we're still on the list, or if we've been removed.
+        Optional<Travel> optionalTravel = PublicTransit.getPlugin().getTravelManager().getTravelFromScheduler(s);
+        if (optionalTravel.isEmpty()) {
+            s.cancel(); // double tapping
+        } else {
```
Comment (mosemister) [inline src/main/java/org/colonelkai/publictransit/TravelManager.java]: personally prefer not to use ``else`` where possible. try

```java
if(optionalTravel.isPresent()){
  optionalTravel.get().travelToNext();
  return;
}
scheduler.cancel();
```

File: src/main/java/org/colonelkai/publictransit/TravelManager.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Consumer;
+
+public class TravelManager {
+    private final HashMap<Travel, Scheduler> travelMap = new HashMap<>();
+
+    private final Consumer<Scheduler> consumer = (Scheduler s) -> {
```
Comment (mosemister) [inline src/main/java/org/colonelkai/publictransit/TravelManager.java]: Single character names is bad practise in any language. Please use ``scheduler`` or some other full word(s) to describe what it is. Exception being if you dont use it

File: src/main/java/org/colonelkai/publictransit/TravelManager.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    terminates if so, not letting the runAfter schedule another scheduler to schedule. what? yeah.
+
+    something like that.
+     */
+
+    private Scheduler scheduleNextTravel(Travel travel) {
+        return TranslateCore.getScheduleManager().schedule()
+                .setDelay(travel.getCurrentNode().getTime())
+                .setDelayUnit(TimeUnit.SECONDS)
+                // RUNNER
```
Comment (mosemister) [inline src/main/java/org/colonelkai/publictransit/TravelManager.java]: Is this comment really needed

hey look, this is the runner which is called with ``setRunner`` 
oh s*** really? couldnt have guessed

File: src/main/java/org/colonelkai/publictransit/TravelManager.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+                .setRunner(
+                        (Scheduler scheduler) -> {
+                            // check if we're still on the list, or if we've been removed.
+                            Optional<Travel> optionalTravel = PublicTransit.getPlugin().getTravelManager().getTravelFromScheduler(scheduler);
+                            if (optionalTravel.isEmpty()) {
+                                scheduler.cancel(); // double tapping
+                                return;
+                            }
+                                optionalTravel.get().travelToNext();
```
Comment (mosemister) [inline src/main/java/org/colonelkai/publictransit/TravelManager.java]: this line seems to want to stand out of line. Any reason?

File: src/main/java/org/colonelkai/publictransit/TravelManager.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+                                = TranslateCore.getServer().getOnlinePlayers()
+                                .parallelStream()
+                                .filter(p -> p.getUniqueId().equals(travel.getPlayer()))
+                                .findAny();
+                        if(optionalLivePlayer.isEmpty()) {
+                            return;
+                        }
+                        LivePlayer player = optionalLivePlayer.get();
+
+                        player.setPosition(travel.getStartingNode().getPosition());
```
Comment (mosemister) [inline src/main/java/org/colonelkai/publictransit/TravelManager.java]: From my understand of this, if the player disconnects, you then schedule every 5 minecraft ticks to check if they have rejoined and if so then teleport them back to the travel start?

Wouldnt it be more logical to set the position on player join back to the starting position? ~~or better yet, set the offline player location (I know its not in the translation layer .... ill check that)~~ (paper doesnt support setting position of offline players .... why????)

File: src/main/java/org/colonelkai/publictransit/TravelManager.java (hunk 6 above)
Comment (ColonelKai) [reply src/main/java/org/colonelkai/publictransit/TravelManager.java]: I've added them to clearly demarkate both Lambda Functions, to see them separately clearly. It's for my dum brain.

File: src/main/java/org/colonelkai/publictransit/TravelManager.java (hunk 8 above)
Comment (ColonelKai) [reply src/main/java/org/colonelkai/publictransit/TravelManager.java]: > or better yet, set the offline player location (I know its not in the translation layer .... ill check that) (paper doesnt support setting position of offline players .... why????)

I didn't know that was an option, and I guess it's not 😆 

> Wouldnt it be more logical to set the position on player join back to the starting position?

ooh yeah, I'll do this instead in a moment.

Comment (mosemister) [reply src/main/java/org/colonelkai/publictransit/TravelManager.java]: The players location will be stored somewhere (likely player data in the world folder) so you should be able to modify it. Sponge actually allows the manipulation of the position and world (as a resource key). Just paper doesn't

File: src/main/java/org/colonelkai/publictransit/utils/PlayerReturner.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    @HEvent
+    public void onPlayerJoinEvent(ClientConnectionEvent.Incoming.Joined event) {
+        LivePlayer player = event.getEntity();
+
+        playersToBeReturned
+                .forEach((uuid, position) -> {
+                    if(player.getUniqueId().equals(uuid)) {
+                        player.setPosition(position);
```
Comment (mosemister) [inline src/main/java/org/colonelkai/publictransit/utils/PlayerReturner.java]: try something like

```java
LivePlayer player = event.getEntity();

if (playersToBeReturned.containsKey(player.getUniqueId())){
  player.setPosition(playersToBeReturned.get(player.getUniqueId()));
  playersToBeReturned.remove(player.getUniqueId())
}
```

This prevents the player teleporting back to this location if they disconnect and reconnect

**Comment by ColonelKai** [comment]
bippity boppity this pr looks fine to me

I totally read it

**Pull request merged**